src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py

```diff
@@ -205,16 +205,13 @@ def __call__(
         # However this currently doesn't work in `mps`.
         latents_device = "cpu" if self.device.type == "mps" else self.device
         latents_shape = (batch_size, self.unet.in_channels, height // 8, width // 8)
+        latents_dtype = text_embeddings.dtype
         if latents is None:
-            latents = torch.randn(
-                latents_shape,
-                generator=generator,
-                device=latents_device,
-            )
+            latents = torch.randn(latents_shape, generator=generator, device=latents_device, dtype=latents_dtype)
         else:
             if latents.shape != latents_shape:
                 raise ValueError(f"Unexpected latents shape, got {latents.shape}, expected {latents_shape}")
-        latents = latents.to(self.device)
+            latents = latents.to(self.device)
```

Comment:
I think this needs to stay outside since `self.device` is not necessarily equal to `latents_device`

Reply:
But then we are moving the tensors twice.

if one is correct, then we should use the correct one directly (either `self.device` or `latents_device`).

You think `self.device` is better than `latents_device`?

Reply:
> Ok ,correct me if I'm wrong, but I think latents_device only exists because randn does not exist for mps. So we need to generate the noise on CPU and then move it to mps.

That is correct.

However, if the user does not provide their own latents, we need to generate them in the CPU and then move them to the `mps` device which if I understand it right would not be happening after this change.

Is `.to(device)` still blocking if no move is necessary?

Reply:
Yes it's blocking even if nothing happens (iirc).

If you check the `mps` block, we already moved to `self.device` for them so this line should really not be necessary.

Reply:
> If you check the `mps` block, we already moved to `self.device` for them so this line should really not be necessary.
👍 

The difference then is that if the user provides their own latents they need to be in the right device. That's fine with me!